### PR TITLE
feat: implement signer metrics exporter

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/1049.

This PR adds some metrics around the following:
* New bitcoin block notification from bitcoin-core.
* New stacks block webhook from stacks-core.
* Stacks and bitcoin validation, with timings.
* Signing round duration with a counter.
* Transaction submission.

## Changes

* Instrument the application with some metrics.
* Add a configurable parameter to enable a Prometheus scrape exporter.

## Testing Information

I will add local prometheus setup in a separate PR.

## Checklist:

- [x] I have performed a self-review of my code

### CODE REVIEW

File: signer/src/transaction_coordinator.rs [hunk 1]
```diff
@@ -540,6 +541,13 @@ where
             )
             .await?;
 
+            metrics::counter!(
```
Comment (matteojug) [inline signer/src/transaction_coordinator.rs:544]: nit: should we also add a counter for the number of deposits swept?

**Review by matteojug — APPROVED** [review]
Nice!

File: signer/src/api/new_block.rs [hunk 2]
```diff
@@ -79,6 +79,12 @@ enum UpdateResult {
 #[tracing::instrument(skip_all, name = "new-block")]
 pub async fn new_block_handler(state: State<ApiState<impl Context>>, body: String) -> StatusCode {
     tracing::debug!("received a new block event from stacks-core");
+    metrics::counter!(
+        "blocks_observed_total",
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/api/new_block.rs:83]: I'm not sure if we could, but let me check (sometimes these macros are picky with their literals).

File: signer/src/lib.rs [hunk 3]
```diff
@@ -76,3 +77,16 @@ pub const MAX_REORG_BLOCK_COUNT: i64 = 10;
 /// The maximum number of sweep transactions that the signers can confirm
 /// per block.
 pub const MAX_TX_PER_BITCOIN_BLOCK: i64 = 25;
+
+/// These are all build info variables. Many of them are set in build.rs.
+
+/// The name of the binary that is being run,
+pub const PACKAGE_NAME: &str = env!("CARGO_PKG_NAME");
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/lib.rs:84]: Yeah sure, good idea.

File: signer/src/transaction_coordinator.rs [hunk 4]
```diff
@@ -540,6 +541,13 @@ where
...
+                "transactions_submitted_total",
+                "blockchain" => "stacks",
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs:546]: oops

File: signer/src/transaction_coordinator.rs [hunk 5]
```diff
@@ -703,10 +723,26 @@ where
...
         let tx = self
             .sign_stacks_transaction(sign_request, multi_tx, chain_tip, wallet)
             .await?;
 
+        metrics::histogram!(
+            "signing_round_duration_seconds",
+            "blockchain" => "stacks",
+            "kind" => kind,
+        )
+        .record(instant.elapsed());
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs:738]: Hmmm well I didn't think it would be that interesting.

[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs:738]: But I'm not opposed. I'll take another look.

File: signer/src/api/new_block.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/api/new_block.rs:83]: We can! I will update all of them.

File: signer/src/transaction_signer.rs [hunk 6]
```diff
@@ -649,7 +678,25 @@ where
...
+                    Err(Error::InvalidSigHash(_)) => "invalid-sighash",
+                    Err(_) => "unexpected-failure",
+                };
+
+                metrics::counter!(
+                    "transaction_validation_total",
+                    "blockchain" => "bitcoin",
+                    "status" => validation_status,
+                )
+                .increment(1);
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_signer.rs:697]: Oh yeah, probably 👍🏽.

File: signer/src/transaction_coordinator.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Yeah it's definitely interesting, will add

File: signer/src/transaction_coordinator.rs [hunk 7]
```diff
@@ -540,6 +541,13 @@ where
             )
             .await?;
 
+            metrics::counter!(
+                crate::metrics::TRANSACTIONS_SUBMITTED_TOTAL,
+                "blockchain" => "bitcoin",
+                "status" => "success",
+            )
+            .increment(1);
```
Comment (matteojug) [inline signer/src/transaction_coordinator.rs:549]: nit: we already increment this in `sign_and_broadcast`.

Comment (djordon) [reply signer/src/transaction_coordinator.rs]: 🙏🏽

File: signer/src/transaction_coordinator.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Added in https://github.com/stacks-network/sbtc/pull/1084/commits/64ee3c2bbd290c800325acaed5851fa095fdeab7

File: signer/src/transaction_coordinator.rs (hunk 1 above)
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Done in https://github.com/stacks-network/sbtc/pull/1084/commits/10b563c5e15aa361b2a56e2857c7d69e5d4e7e42

File: signer/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/lib.rs:84]: Done in https://github.com/stacks-network/sbtc/pull/1084/commits/64ee3c2bbd290c800325acaed5851fa095fdeab7

File: signer/src/main.rs [hunk 8]
```diff
@@ -69,7 +69,15 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     // Configure the binary's stdout/err output based on the provided output format.
     let pretty = matches!(args.output_format, Some(LogOutputFormat::Pretty));
-    signer::logging::setup_logging("", pretty);
+    signer::logging::setup_logging("info,signer=debug", pretty);
```
Comment (matteojug) [inline signer/src/main.rs:72]: If we do this here does it mean we will override whatever it's set from outside? eg, in docker-compose

File: signer/src/api/new_block.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/api/new_block.rs]: Updated across many commits.

File: signer/src/main.rs (hunk 8 above)
Comment (djordon) [reply signer/src/main.rs:72]: yeah, the environment variable takes precedence. 

https://github.com/stacks-network/sbtc/blob/845f6dd0b9dfed81afebdab04061bd6b1241cfe4/signer/src/logging.rs#L22-L37

**Pull request merged**